internal/websocket/ws.go

```diff
@@ -284,7 +289,21 @@ func (ws *Websocket) run(ctx context.Context) {
 			// Read loop
 			for {
 				var msg wrp.Message
+				ctx, cancel := context.WithTimeout(ctx, ws.inactivityTimeout)
 				typ, reader, err := ws.conn.Reader(ctx)
+				if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
Don't forget that both `context.DeadlineExceeded` and `context.Canceled` can be returned by the `context` package.  `context.Canceled` means that something called the cancel function before the timeout occurred, which would indicate that something external to this type requested shutdown of the `Websocket`.  That's may be a slightly different case, I'm not sure.

Reply:
good point! I think we're ok because our ending select will catch the case when our parent ctx has been cancelled

```go
select {
case <-time.After(next):
case <-ctx.Done():
	return
}
```

Reply:
Those are still (2) cases though.  A unit test should cover both, and verify that both of those channels being signaled produces the desired behavior in each case.

Reply:
💯 just added the test `internal/websocket/ws_test.go: Test_CancelCtx` to cover this case 🙂